File: uspex/read_poscar.m
% this function is usedt ot read POSCAR files .
% input is the input file's name .
function [coor,LATT,numIons,natom,atom_chan] = read_poscar(input)
[fid,message] = fopen(input);

tmp = fgetl(fid) ;    % tile .
factor = str2num(fgetl(fid)) ; % the zoom factor .

LATT =zeros(3);    % the bulk LATTIC£®
for i =1 : 3
    tmp = fgetl(fid);
    LATT(i,:) = str2num(tmp);
end
LATT = LATT*factor;

% atom type
tmp = fgetl(fid);
ntyp = str2num(tmp); % number of each type atom .
numIons = ntyp;

if isempty(ntyp)
    tmp= fgetl(fid);
    ntyp = str2num(tmp);
    numIons = ntyp;
end
% direct
natom = sum(ntyp);
tmp = fgetl(fid);  % the line is the string 'Direct'
if findstr(tmp,'S')
	tmp = fgetl(fid);
end

coor = zeros(natom,3);   %¡¡atom 's coordination .
atom_chan = zeros(natom,3);  % the atom fix or not , if atom_chan(ii,1) ==1 , it means that the x dirction is not fix .

for ii = 1 : natom
    tmp=fgetl(fid);
    temp = regexp(tmp,'\s+','split');
    if isempty(temp{1})
        temp=temp(2:end);
    end
    for jj = 1 : 3
        coor(ii,jj)=str2num(temp{jj});
        try
            if temp{jj+3} == 'T'
                atom_chan(ii,jj) = 1;
            else
                atom_chan(ii,jj) = 0 ;
            end
        catch
        end
    end
end
fclose(fid);
end

File: uspex/read_poscar.py
import numpy as np


def read_poscar(poscat_file_path):
    with open(poscat_file_path) as file_poscar:
        title = file_poscar.readline().strip()  # useless

        # get the first line, usually called name
        factor = float(file_poscar.readline().strip())
        lattice = np.zeros((3, 3))
        for i in range(3):
            tem_line = file_poscar.readline().strip().split(' ')
            tem_line = filter_blank(tem_line)
            lattice[i, :] = np.array(tem_line, dtype=float)
        lattice = lattice * factor

        # get the lattice parameters
        atom_type = file_poscar.readline().strip().split('    ')
        # return a list containing Atom type
        atom_number = file_poscar.readline().strip().split('    ')
        atom = {}

        # return a dictoion key=atom type, value=atom number
        for i in range(len(atom_type)):
            atom[atom_type[i]] = int(atom_number[i])
        total_number = sum(atom.values())
        coordinate = np.zeros((total_number, 4))

        # jump two lines (null)
        for i in range(2):
            _ = file_poscar.readline()

        # read coordinate data
        # coordinate(x, y, z)  (fix or unfix)
        for i in range(total_number):
            temp = file_poscar.readline()
            tt = temp.strip().split(' ')
            tt = filter_blank(tt)
            coordinate[i, 0:3] = np.array(tt[0:3], dtype=float)
            if 'T' in tt:
                # 1 means T,  0 means F
                coordinate[i, -1] = 1
            else:
                coordinate[i, -1] = 0
    return lattice, atom, coordinate


def filter_blank(line):
    '''
    filter blank in  a given list
    '''
    return [_ for _ in line if _]


path = '/Users/zhangjiawei/Code/zjw/uspex/POSCAR'
lattice, atom, coordinate = read_poscar(path)

print lattice
print atom
print coordinate
